modules/oe_theme_content_event/src/Plugin/ExtraField/Display/ProgrammeExtraField.php

```diff
@@ -0,0 +1,219 @@
+    parent::__construct($configuration, $plugin_id, $plugin_definition, $entity_type_manager);
+    $this->dateFormatter = $date_formatter;
+    $this->entityRepository = $entity_repository;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('entity_type.manager'),
+      $container->get('date.formatter'),
+      $container->get('entity.repository')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getLabel() {
+    return $this->t('Programme');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function viewElements(ContentEntityInterface $entity) {
+    if ($entity->get('oe_event_programme')->isEmpty()) {
+      return;
+    }
+
+    $cache = new CacheableMetadata();
+
+    $build = [
+      '#type' => 'pattern',
+      '#id' => 'timeline',
+      '#fields' => [
+        'limit' => 5,
+        'title' => $this->getLabel(),
+        'items' => [],
+      ],
+    ];
+
+    // Retrieve array of Programme item entities.
+    $programmes = $entity->get('oe_event_programme')->referencedEntities();
+
+    // Sort array of event programme items by start date's timestamp.
+    usort($programmes, function ($a, $b) {
+      if ($a->get('oe_event_programme_dates')->start_date->getTimestamp() === $b->get('oe_event_programme_dates')->start_date->getTimestamp()) {
+        return 0;
+      }
+      return $a->get('oe_event_programme_dates')->start_date->getTimestamp() < $b->get('oe_event_programme_dates')->start_date->getTimestamp() ? -1 : 1;
+    });
+
+    $items = [];
+    $previous_end_date = NULL;
+    foreach ($programmes as $programme) {
+      // Skip invalid items.
+      if (!$programme instanceof ProgrammeItemInterface || $programme->get('oe_event_programme_dates')->isEmpty()) {
+        continue;
+      }
+
+      $programme = $this->entityRepository->getTranslationFromContext($programme);
+      $cache->addCacheableDependency($programme);
+
+      $item = [
+        'label' => $this->generateEventProgrammeLabel($programme, $previous_end_date),
+        'title' => $programme->get('name')->getString(),
+      ];
+      if (!$programme->get('oe_description')->isEmpty()) {
+        $item['body'] = [
+          '#type' => 'processed_text',
+          '#text' => $programme->get('oe_description')->first()->getValue()['value'],
+          '#format' => $programme->get('oe_description')->first()->getValue()['format'],
+          '#langcode' => $entity->language(),
+        ];
+      }
+      $items[] = $item;
+
+      // Keep day of previous event programme item for possible comparison.
+      $previous_end_date = $this->dateFormatter->format($programme->get('oe_event_programme_dates')->end_date->getTimestamp(), 'custom', 'Ymd');
+    }
+
+    if (!$items) {
+      return [];
```

Comment:
What about all the cache data we accumulated above?

Reply:
We didn't collect any cache metadata because didn't have any reference entities.